sycl/include/sycl/ext/intel/experimental/esimd/math.hpp

```diff
@@ -1592,39 +1615,69 @@ asin(T src0, int flag = saturation_off) {
   return Result[0];
 }
 
+namespace detail {
+// std::numbers::ln2_v<float> in c++20
+constexpr float ln2 = 0.69314718f;
+// std::numbers::log2e_v<float> in c++20
+constexpr float log2e = 1.442695f;
+} // namespace detail
+
 /// Computes the natural logarithm of the given argument. This is an
 /// emulated version based on the H/W supported log2.
 /// @param the source operand to compute base-e logarithm of.
 /// @return the base-e logarithm of \p src0.
-template <int SZ>
-ESIMD_NODEBUG ESIMD_INLINE simd<float, SZ> log(simd<float, SZ> src0,
-                                               int flag = saturation_off) {
-  constexpr float ln2 = 0.69314718f; // std::numbers::ln2_v<float> in c++20
-  simd<float, SZ> Result = esimd::log2<SZ>(src0) * ln2;
+template <class T, int SZ>
+ESIMD_NODEBUG ESIMD_INLINE simd<T, SZ> log(simd<T, SZ> src0, int flag) {
+  using CppT = __SEIEED::__cpp_t<T>;
+  simd<T, SZ> Result = esimd::log2<T, SZ, saturation_off>(src0) * detail::ln2;
 
   if (flag != saturation_on)
     return Result;
 
-  return esimd::saturate<float>(Result);
+  return esimd::saturate<T>(Result);
 }
 
-ESIMD_NODEBUG ESIMD_INLINE float log(float src0, int flag = saturation_off) {
-  return esimd::log<1>(src0, flag)[0];
+template <class T, int SZ, int Flag = saturation_off>
+ESIMD_NODEBUG ESIMD_INLINE simd<T, SZ> log(simd<T, SZ> src0) {
+  simd<T, SZ> Result = esimd::log2<T, SZ, saturation_off>(src0) * detail::ln2;
+
+  if constexpr (Flag != saturation_on)
+    return Result;
+
+  return esimd::saturate<T>(Result);
```

Comment:
I remember there was an interesting detail about 'if constexpr', could not quickly find a useful link this time though.
Code after  'if constexpr (cond) return smth;' is not automatically handled as if it was under 'if constexpr (!cond).
In general, for such cases it is better to add 'else' branch explicitly.
```suggestion
  if constexpr (Flag != saturation_on)
    return Result;
  else
    return esimd::saturate<T>(Result);
```

Reply:
Probably you meant something like [this](https://stackoverflow.com/questions/53818624/why-does-if-constexpr-require-an-else-to-work).
This relates to a case when what comes after if constexpr (wrapped under else in your suggestion) does not pass semantics check and leads to compilation error. This is not the case in this case.
Anyway, adding else might be clearer for some readers - will add.